packages/lib/src/operations.ts

```diff
@@ -20,6 +20,164 @@ export const didForCreateOp = async (op: t.CompatibleOp, truncate = 24) => {
+  signingKey: string
+  handle: string
+  pds: string
+  rotationKeys: string[]
+  prev: CID | null
+}): t.UnsignedOperation => {
+  return {
+    type: 'plc_operation',
+    verificationMethods: {
+      atproto: opts.signingKey,
+    },
+    rotationKeys: opts.rotationKeys,
+    alsoKnownAs: [ensureAtprotoPrefix(opts.handle)],
+    services: {
+      atproto_pds: {
+        type: 'AtprotoPersonalDataServer',
+        endpoint: ensureHttpPrefix(opts.pds),
+      },
+    },
+    prev: opts.prev?.toString() ?? null,
+  }
+}
+
+export const atprotoOp = async (opts: {
+  signingKey: string
+  handle: string
+  pds: string
+  rotationKeys: string[]
+  prev: CID | null
+  signer: Keypair
+}) => {
+  return addSignature(formatAtprotoOp(opts), opts.signer)
+}
+
+export const createOp = async (opts: {
+  signingKey: string
+  handle: string
+  pds: string
+  rotationKeys: string[]
+  signer: Keypair
+}): Promise<{ op: t.Operation; did: string }> => {
+  const op = await atprotoOp({ ...opts, prev: null })
+  const did = await didForCreateOp(op)
+  return { op, did }
+}
+
+export const createUpdateOp = async (
+  lastOp: t.CompatibleOp,
+  signer: Keypair,
+  fn: (normalized: t.UnsignedOperation) => Omit<t.UnsignedOperation, 'prev'>,
+): Promise<t.Operation> => {
+  const prev = await cidForCbor(lastOp)
+  // omit sig so it doesn't accidentally make its way into the next operation
+  const { sig, ...normalized } = normalizeOp(lastOp)
+  const unsigned = await fn(normalized)
+  return addSignature(
+    {
+      ...unsigned,
+      prev: prev.toString(),
+    },
+    signer,
+  )
+}
+
+export const updateAtprotoKeyOp = async (
+  lastOp: t.CompatibleOp,
+  signer: Keypair,
+  atprotoKey: string,
+): Promise<t.Operation> => {
+  return createUpdateOp(lastOp, signer, (normalized) => ({
+    ...normalized,
+    verificationMethods: {
+      ...normalized.verificationMethods,
+      atproto: atprotoKey,
+    },
+  }))
+}
+
+export const updateHandleOp = async (
+  lastOp: t.CompatibleOp,
+  signer: Keypair,
+  handle: string,
+): Promise<t.Operation> => {
+  const formatted = ensureAtprotoPrefix(handle)
+  return createUpdateOp(lastOp, signer, (normalized) => {
+    const handleI = normalized.alsoKnownAs.findIndex((h) =>
+      h.startsWith('at://'),
+    )
+    let aka: string[]
+    if (handleI < 0) {
+      aka = [formatted, ...normalized.alsoKnownAs]
+    } else {
+      aka = [
+        ...normalized.alsoKnownAs.slice(0, handleI),
+        formatted,
+        ...normalized.alsoKnownAs.slice(handleI + 1),
+      ]
+    }
+    return {
+      ...normalized,
+      alsoKnownAs: aka,
+    }
+  })
+}
+
+export const updatePdsOp = async (
+  lastOp: t.CompatibleOp,
+  signer: Keypair,
+  endpoint: string,
+): Promise<t.Operation> => {
+  const formatted = ensureHttpPrefix(endpoint)
+  return createUpdateOp(lastOp, signer, (normalized) => {
+    return {
+      ...normalized,
+      services: {
+        ...normalized.services,
+        atproto_pds: {
```

Comment:
Just curious, where does the snake-casing convention stem from in this context?

Reply:
for two reasons:
- the keys of this map become the `id` in the context of the did doc, which i believe are meant to be case-insensitive: https://www.w3.org/TR/did-core/#services
- but you can't write record keys in kebab-case (at least nicely, you can wrap them in quotes i guess) so I used snake_case instead

Reply:
Right on, thanks!